### DISCUSSION TITLE
how to make the server non-blocking

### DISCUSSION THREAD

**Discussion by tianleizhang** [cesanta/mongoose#2661] [Ask questions, get answers] [answered]
Hi, I am new to mongoose web server. I just imported the mongoose into my project and made it work. One thing I noticed is that the server works in a single thread and everything is driven by mg_mgr_poll in the the main loop context. If there is a HTTP request and the server cannot get the response ready quickly (in our case the server is just a local network gateway and it needs to collect the related information from other devices in the local network before it can compose the HTTP response), I suppose the service will be blocked, i.e., the server won't able to respond to other requests before the callback fn returns. Is my understanding on this right? If so, if there a way to avoid this?

**Comment by scaprile** [comment, marked as answer]
Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.

From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/).

**Reply by tianleizhang** [reply]
Thank you so much Scaprile! this is exactly what I am looking for. You answer helped me a lot.
